### PR TITLE
Better handling of HTTP errors in `PocoHTTPClient`

### LINKED ISSUE #23775: S3 table endpoint region configuration constrained to the amazonaws domain
Hi, first of all thank you for a great piece of software.

**Describe the bug**

PR  #15646 fixes #10417 by matching the S3 endpoint region using the following regular expression: `s3.<region>.amazonaws`. Attempting to use an S3 compatible storage engine hosted outside of AWS an authentication error due to mismatch with the default region (`us-east-1`).

**Does it reproduce on recent release?**
Reproduces on `ClickHouse server version 21.4.5.46 (official build)`

**How to reproduce?**

The following query fails:

`SELECT * from s3('https://s3.gra.cloud.ovh.net/..., 'key', 'secret', 'csv', '...') LIMIT 10`

**Expected behavior**

One potential solution would be to allow the region to be specified in the S3 disk settings (adding vendor-specific regexp patterns in  `PocoHTTPClientConfiguration::updateSchemeAndRegion` seems unrealistic).

### PR DESCRIPTION
**Pull request by excitoon**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Better handling of HTTP errors in `PocoHTTPClient`. Response bodies of HTTP errors were being ignored earlier.

Fixes #23775

### CODE REVIEW

**Comment by excitoon** [comment]
@Jokser Can you please look at my changes?

**Review by alexey-milovidov — APPROVED** [review]

**Comment by Jokser** [comment]
So, now any positive status code (200) will be treated as error?

**Comment by excitoon** [comment]
> So, now any positive status code (200) will be treated as error?

@Jokser No, thing is AWS client handles HTTP errors itself (by response code), so we shall not report `Client` type of error in these cases.

**Comment by Jokser** [comment]
@excitoon I mean metrics increment `ProfileEvents::increment(select_metric(S3MetricType::Errors));`

**Comment by excitoon** [comment]
@Jokser I see, thank you, this is a typo.

**Comment by excitoon** [comment]
Failed tests are unrelated.

File: src/IO/S3/PocoHTTPClient.cpp [hunk 1]
```diff
@@ -254,29 +254,16 @@ void PocoHTTPClient::makeRequestInternal(
...
             {
-                String error_message;
-                Poco::StreamCopier::copyToString(response_body_stream, error_message);
-
-                if (Aws::Http::IsRetryableHttpResponseCode(response->GetResponseCode()))
-                    response->SetClientErrorType(Aws::Client::CoreErrors::NETWORK_CONNECTION);
-                else
-                    response->SetClientErrorType(Aws::Client::CoreErrors::USER_CANCELLED);
-
-                response->SetClientErrorMessage(error_message);
```
Comment (vdimir) [inline src/IO/S3/PocoHTTPClient.cpp:267]: Why we don't set `SetClientErrorMessage` and `SetClientErrorType`, but only `SetResponseBody`. How does client that reads response will get that it's an error?

Comment (excitoon) [reply src/IO/S3/PocoHTTPClient.cpp:267]: By response code. Client error is not HTTP error but rather a connection error/SSL, logical errors and so on.

**Comment by excitoon** [comment]
Rebased to master.

File: src/IO/S3/PocoHTTPClient.cpp (hunk 1 above)
Comment (vdimir) [reply src/IO/S3/PocoHTTPClient.cpp:302]: Ok, we set `response->SetResponseCode` above got it.

**Review by vdimir — APPROVED** [review]

**Pull request merged**